File: lib/src/download.ts
import { GetObjectCommand, type S3Client } from "@aws-sdk/client-s3";
import { createHash } from "node:crypto";
import { createWriteStream } from "node:fs";
import { Readable, Transform } from "node:stream";
import { pipeline } from "node:stream/promises";

export interface DownloadResult {
  sha256: string;
  bytes: number;
}

/**
 * Streams an S3 object to disk, computing SHA-256 along the way. Saves a
 * second full read of the source for the byte-hash dedup check.
 */
export async function downloadAndHash(
  client: S3Client,
  bucket: string,
  key: string,
  localPath: string,
): Promise<DownloadResult> {
  const res = await client.send(new GetObjectCommand({ Bucket: bucket, Key: key }));
  if (!res.Body) throw new Error(`Empty body for s3://${bucket}/${key}`);

  const stream = res.Body as Readable;
  const hash = createHash("sha256");
  let bytes = 0;

  const teeAndHash = new Transform({
    transform(chunk, _enc, cb) {
      const buf: Buffer = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
      hash.update(buf);
      bytes += buf.length;
      cb(null, buf);
    },
  });

  await pipeline(stream, teeAndHash, createWriteStream(localPath));

  return { sha256: hash.digest("hex"), bytes };
}

File: core/download.go
package core

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"os"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

// DownloadResult is the outcome of a stream-and-hash download.
type DownloadResult struct {
	SHA256 string
	Bytes  int64
}

// DownloadAndHash streams an S3 object to localPath while computing its SHA-256,
// saving a second full read for the byte-hash dedup check.
func DownloadAndHash(ctx context.Context, client *s3.Client, bucket, key, localPath string) (DownloadResult, error) {
	out, err := client.GetObject(ctx, &s3.GetObjectInput{Bucket: aws.String(bucket), Key: aws.String(key)})
	if err != nil {
		return DownloadResult{}, err
	}
	defer out.Body.Close()

	f, err := os.Create(localPath)
	if err != nil {
		return DownloadResult{}, err
	}
	defer f.Close()

	h := sha256.New()
	n, err := io.Copy(io.MultiWriter(f, h), out.Body)
	if err != nil {
		return DownloadResult{}, fmt.Errorf("streaming s3://%s/%s: %w", bucket, key, err)
	}
	return DownloadResult{SHA256: hex.EncodeToString(h.Sum(nil)), Bytes: n}, nil
}
